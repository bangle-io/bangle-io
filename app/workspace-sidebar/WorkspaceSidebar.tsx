import React from 'react';

import type { SidebarType } from '@bangle.io/extension-registry';
import {
  Button,
  ChevronLeftIcon,
  CloseIcon,
  ErrorBoundary,
} from '@bangle.io/ui-components';

export function WorkspaceSidebar({
  onDismiss,
  sidebar,
  widescreen,
}: {
  onDismiss: () => void;
  sidebar: SidebarType;
  widescreen: boolean;
}) {
  return (
    <div
      data-testid="app-workspace-sidebar_workspace-sidebar"
      className="flex flex-col flex-grow h-full smallscreen:min-h-screen overflow-y-scroll bg-colorNeutralBgLayerTop"
    >
      <div className="flex flex-row justify-between px-2 mt-2">
        <span className="font-bold self-center">{sidebar.title}</span>
        <span>
          <Button
            variant="transparent"
            onPress={onDismiss}
            size="sm"
            ariaLabel={'Hide ' + sidebar.title}
            tooltipPlacement="bottom"
            leftIcon={widescreen ? <ChevronLeftIcon /> : <CloseIcon />}
          />
        </span>
      </div>
      <ErrorBoundary>
        <sidebar.ReactComponent />
      </ErrorBoundary>
    </div>
  );
}
