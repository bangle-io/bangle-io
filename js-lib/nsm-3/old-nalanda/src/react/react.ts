import { useDebugValue, useRef, useState } from 'react';
import useSyncExternalStoreExports from 'use-sync-external-store/shim';

import type { Slice } from '../vanilla/slice';
import type { Store } from '../vanilla/store';
import type { AnySlice } from '../vanilla/types';

const { useSyncExternalStore } = useSyncExternalStoreExports;

type InferData<T> = T extends Slice<any, infer TData, any> ? TData : never;

interface ReactAdapter<TSnapshot> {
  subscribe: (onStoreChange: () => void) => () => void;
  getSnapshot: () => TSnapshot;
}

export function createUseTrackSliceHook<TAllSliceName extends string = any>(
  store: Store<TAllSliceName>,
) {
  function useSlice<TSlice extends AnySlice>(sl: TSlice): InferData<TSlice> {
    const ref = useRef<InferData<TSlice>>();

    const [adapter] = useState(() => {
      const adapter: ReactAdapter<InferData<TSlice>> = {
        subscribe: (onStoreChange) => {
          const sliceEffect = store.effect((effectStore) => {
            const selectedData = sl.track(effectStore);
            ref.current = selectedData;

            onStoreChange();

            queueMicrotask(() => {
              if (effectStore._runInstance?._addTrackedCount === 0) {
                console.warn(
                  `You forgot to destructure/access the tracked field from ${sl.name}. This will not track any changes!.`,
                );
              }
            });
          });

          return () => {
            store.unregisterEffect(sliceEffect);
          };
        },
        getSnapshot: () => {
          return ref.current ?? sl.get(store.state);
        },
      };

      return adapter;
    });

    const data = useSyncExternalStore(adapter.subscribe, adapter.getSnapshot);

    useDebugValue(data);

    return data;
  }

  return useSlice;
}
