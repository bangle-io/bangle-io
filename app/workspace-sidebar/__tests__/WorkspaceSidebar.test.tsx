/**
 * @jest-environment @bangle.io/jsdom-env
 */
import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';

import { WorkspaceSidebar } from '../WorkspaceSidebar';

let originalConsoleError = console.error;

beforeEach(() => {
  console.error = originalConsoleError;
});

test('handles error', () => {
  // silencing the error from polluting the logging
  console.error = jest.fn();
  let result = render(
    <div>
      <WorkspaceSidebar
        widescreen
        onDismiss={jest.fn()}
        sidebar={{
          name: 'sidebar::test-sidebar',
          title: 'search notes',
          activitybarIcon: <span>test-icon</span>,
          ReactComponent: () => {
            throw new Error('Blah blah');

            return <span>something</span>;
          },
          hint: 'test-hint',
        }}
      />
    </div>,
  );
  expect(result.container.innerHTML).toContain('Blah blah');
});

test('renders', () => {
  let result = render(
    <div>
      <WorkspaceSidebar
        widescreen
        onDismiss={jest.fn()}
        sidebar={{
          name: 'sidebar::test-sidebar',
          title: 'search notes',
          activitybarIcon: <span>test-icon</span>,
          ReactComponent: () => {
            return <span>something</span>;
          },
          hint: 'test-hint',
        }}
      />
    </div>,
  );

  expect(result.container.innerHTML).toContain('something');
  expect(result.container).toMatchSnapshot();
});

test('calls the dismiss button', () => {
  const onDismiss = jest.fn();
  let result = render(
    <div>
      <WorkspaceSidebar
        widescreen
        onDismiss={onDismiss}
        sidebar={{
          name: 'sidebar::test-sidebar',
          title: 'search notes',
          activitybarIcon: <span>test-icon</span>,
          ReactComponent: () => {
            return <span>something</span>;
          },
          hint: 'test-hint',
        }}
      />
    </div>,
  );

  act(() => {
    fireEvent.click(result.getByLabelText('Hide search notes'));
  });

  expect(onDismiss).toBeCalledTimes(1);
});
