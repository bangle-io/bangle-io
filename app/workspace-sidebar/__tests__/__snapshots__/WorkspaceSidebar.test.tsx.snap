// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <div>
    <div
      class="flex flex-col flex-grow h-full smallscreen:min-h-screen overflow-y-scroll bg-colorNeutralBgLayerTop"
      data-testid="app-workspace-sidebar_workspace-sidebar"
    >
      <div
        class="flex flex-row justify-between px-2 mt-2"
      >
        <span
          class="font-bold self-center"
        >
          search notes
        </span>
        <span>
          <button
            aria-label="Hide search notes"
            class="text-sm font-600 h-8 min-w-8 px-2  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
            style="background-color: transparent;"
            type="button"
          >
            <span
              class="flex flex-grow-1 overflow-hidden "
              style="justify-content: center;"
            >
              <span>
                <svg
                  class="w-5 h-5  "
                  fill="currentColor"
                  viewBox="0 0 18 18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    fill="#ff13dc"
                    height="18"
                    opacity="0"
                    width="18"
                  />
                  <path
                    d="M6,9a.994.994,0,0,0,.2925.7045l3.9915,3.99a1,1,0,1,0,1.4355-1.386l-.0245-.0245L8.4095,9l3.286-3.285A1,1,0,0,0,10.309,4.28l-.0245.0245L6.293,8.2945A.994.994,0,0,0,6,9Z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </span>
      </div>
      <span>
        something
      </span>
    </div>
  </div>
</div>
`;
